use caddy 2 for https local development

ggilmore opened the pull request:
Fixes: https://github.com/sourcegraph/sourcegraph/issues/1951

This PR adds a new Procfile entry that runs a https reverse proxy at `https://sourcegraph.test:3443` for local dev. 

Note that you'll need to have caddy 2 > 2.0.0-beta.19 installed from now on, see https://github.com/sourcegraph/sourcegraph/blob/caddy-local-https/doc/dev/local_development.md#step-5-configure-https-reverse-proxy for installation and configuration instructions.

---

Note: I tried (for a long time) and failed to get e2e tests working with this. Some combination of web/node/puppeteer doesn't like the self-signed certs and I wasn't able to figure it out. I'll post a WIP PR for @sourcegraph/web to own with my findings after this PR is merged.

uwedeportivo commented on doc/dev/local_development.md:
````diff
@@ -344,6 +421,14 @@ If you ever need to wipe your local database and Redis, run the following comman
 ./dev/drop-entire-local-database-and-redis.sh
 ```
 
+### Caddy certifiate problems
````
```suggestion
### Caddy certificate problems
```

unknwon commented on .tool-versions:
```diff
@@ -1,3 +1,4 @@
 golang 1.14
 yarn 1.22.4
+caddy 2.0.0-beta.17
```
beta 18 was release 9 hours ago, should we use that? https://github.com/caddyserver/caddy/releases/tag/v2.0.0-beta.18

unknwon commented on doc/dev/local_development.md:
```diff
@@ -229,6 +253,59 @@ page](postgresql.md).
 
 Migrations are applied automatically.
 
+### (optional) asdf
```
Assuming new diff is all about asdf, then all headings inside this section should have one more `#`, otherwise they are not visually having subheadings, e.g. `#### Install`.

unknwon commented on doc/dev/local_development.md:
```diff
@@ -4,18 +4,41 @@ Have a look around, our code is on [GitHub](https://sourcegraph.com/github.com/s
-- [How to debug live code](#how-to-debug-live-code)
-- [Windows support](#windows-support)
-- [Other nice things](#other-nice-things)
+- [Getting started with developing Sourcegraph](#getting-started-with-developing-sourcegraph)
```
Is this intentional? Seems the only one bullet point in this indentation level making everything else has one more indentation. 

Also seems a bit weird to have a link point back to H1 at the top of the page.

ggilmore replied:
My VSCode markdown extension automatically keeps things that look like a table of contents to date with the contents of the markdown file. I'll disable it for now.

unknwon commented on doc/dev/local_development.md:
```diff
@@ -4,18 +4,41 @@ Have a look around, our code is on [GitHub](https://sourcegraph.com/github.com/s
-- [Windows support](#windows-support)
-- [Other nice things](#other-nice-things)
+- [Getting started with developing Sourcegraph](#getting-started-with-developing-sourcegraph)
+  - [Outline](#outline)
```
This bullet point also seems not worthing having it?

unknwon commented on doc/dev/local_development.md:
````diff
@@ -344,6 +421,14 @@ If you ever need to wipe your local database and Redis, run the following comman
 ./dev/drop-entire-local-database-and-redis.sh
 ```
 
+### Caddy certificate problems
````
It might be worth pointing out we are using Caddy 2 in all places, since 1 and 2 are quite different in terms of configuration.

unknwon commented on doc/dev/local_development.md:
```diff
@@ -344,6 +421,14 @@ If you ever need to wipe your local database and Redis, run the following comman
 
+### Caddy certificate problems
+
+We use Caddy to setup https for local development. It creates self-signed certificates and uses that to serve the local Sourcegraph instance. If your browser complains about the certificate, check the following:
```
```suggestion
We use Caddy to setup HTTPS for local development. It creates self-signed certificates and uses that to serve the local Sourcegraph instance. If your browser complains about the certificate, check the following:
```

unknwon commented:
Thanks for writing this up! Few comments about formatting.

ggilmore commented:
> Thank you for working on this, this is very valuable to the web team!
> 
> `localhost` sometimes has different behavior in browsers from other domains, so in order to properly test things will work, I'd like to use `https://sourcegraph.test` instead of `https://localhost:3443` (@uwedeportivo's guide has instructions for that). I tried this branch but could not get that to work. Do you know how to get that to work on this branch?

@felixfbecker 
I'm looking into it now, but Caddy won't edit /etc/hosts for you in order to make sourcegraph.test resolve to localhost. You'll need to edit that out of band (and we'd need a solution for the e2e tests as well. Do you still want that specific url? Would something like `https://127.0.0.1:3443` work?

felixfbecker commented:
I already set up `dnsmasq` up following @uwedeportivo's guide, which I think could be something optional for people to use. Maybe we could make the port customizable through an env var?

uwedeportivo commented:
@felixfbecker dnsmasq will resolve .test domains to 127.0.0.1 so https://127.0.0.1:3443 should work.( i uninstalled dnsmasq, sorry so i cannot test. it interfered with multipass and mikrok8s. i somewhat regret recommending it now, sorry. a simple entry in /etc/hosts like @keegancsmith said would have worked too for individual domain names)

uwedeportivo commented:
👍 makes sense

felixfbecker commented:
@uwedeportivo yeah I'm aware :) I just like having a proper domain name, types much faster into a browser extension config box than an IP address, and I don't know if the exceptions that apply to `localhost` also apply to `127.0.0.1`

ggilmore commented:
https://caddy.community/t/caddy-still-trying-to-hit-lets-encrypt-even-with-local-certs-enabled/7268

Currently waiting on the outcome of this to proceed

mholt commented:
My bad. Will get this fixed soon-ish!

creachadair commented:
> This PR adds a new Procfile entry that runs a https reverse proxy at `https://localhost:3443` for local dev. Note that you'll need to have caddy 2 > 2.0.0-beta.17 installed from now on (either manually or though asdf (preferred) - homebrew is not an option).

If it makes your local testing any easier, I added a tap for this: `brew install creachadair/homebrew/caddy@2.0.0-beta.18` should work. I personally find shell alias hacks like asdf to be disruptive and irritating, but YMMV.

mholt commented:
I think I fixed the bug in Caddy, see linked issue.

Also, you can simplify your Caddyfile, this should be all you need:

```
{
	local_certs
}

{$SOURCEGRAPH_HTTPS_DOMAIN}:{$SOURCEGRAPH_HTTPS_PORT} {
	reverse_proxy localhost:3080
}
```

The `https://` scheme is always the default, and `local_certs` now (properly) assumes `tls internal` for all sites, so you don't need to add those to your config separately.

Hope this works better! :sweat_smile:

mholt commented:
@ggilmore @felixfbecker 

> Caddy won't edit /etc/hosts for you in order to make sourcegraph.test resolve to localhost.

Should it? I can look into that. (Wouldn't hold your breath for this, though, might not get done until v2.1 or something.)

ggilmore commented:
> @ggilmore @felixfbecker
> 
> > Caddy won't edit /etc/hosts for you in order to make sourcegraph.test resolve to localhost.
> 
> Should it? I can look into that. (Wouldn't hold your breath for this, though, might not get done until v2.1 or something.)

Thanks @mholt for the quick fix! 

It would be really nice to be able to use a single program to be able to set up my local environment for applications (like ours) that need to be tested via HTTPS. However, I'd find it surprising for any app to silently modify /etc/hosts (e.g. in this case I suddenly find an entry for sourcegraph.test pointing to localhost). 

Maybe if it was an explicit flag / Caddyfile directive?

mholt commented:
> Maybe if it was an explicit flag / Caddyfile directive?

Yeah, definitely wouldn't be implicit (esp. since it requires root privileges!). I'll add it to the list, then. If it's something you want to see sooner rather than later, feel free to open an issue upstream -- but again, it will probably be after the 2.0 release (which isn't too far off).

ggilmore commented:
@sourcegraph/distribution @sourcegraph/web @unknwon PTAL and test my documentation, I updated and rebased my PR with the caddy fixes and other PR comments

uwedeportivo commented:
LGTM

beyang commented:
Could we auto-install Caddy like we do for Jaeger and docsite given that it is a single Go binary? E.g.,
* https://github.com/sourcegraph/sourcegraph/blob/master/dev/jaeger.sh#L13-L18
* https://github.com/sourcegraph/sourcegraph/blob/master/dev/docsite.sh#L13-L16

felixfbecker commented on dev/Procfile:
```diff
@@ -8,6 +8,7 @@ github-proxy: github-proxy
 frontend: env CONFIGURATION_MODE=server SITE_CONFIG_ESCAPE_HATCH_PATH=$HOME/.sourcegraph/site-config.json frontend
 watch: ./dev/changewatch.sh
 nginx: nginx -p . -g 'daemon off;' -c $PWD/dev/nginx.conf 2>&1 | grep -v 'could not open error log file'
+caddy: env SOURCEGRAPH_HTTPS_DOMAIN=sourcegraph.test SOURCEGRAPH_HTTPS_PORT=3443 caddy run --watch --config=dev/Caddyfile
```
Could we set the env vars in `./dev/start.sh` instead as default values, so that it's still possible to override them?

ggilmore commented:
@felixfbecker @beyang PTAL, updated

unknwon commented on dev/add_https_domain_to_hosts.sh:
```diff
@@ -0,0 +1,26 @@
+fi
+
+
+SOURCEGRAPH_HTTPS_DOMAIN="${SOURCEGRAPH_HTTPS_DOMAIN:-"sourcegraph.test"}"
```
nit: what about "sourcegraph.local"? Just feels more nature as we are going to use this script mostly for local development.

ggilmore replied:
https://news.ycombinator.com/item?id=12579128

You can also change it via the env vars if you still don’t like the domain

unknwon replied:
haha, thanks for the link!

unknwon commented on doc/dev/local_development.md:
```diff
@@ -173,6 +176,54 @@ The following are two recommendations for installing these dependencies:
+
+We use asdf in buildkite to lock the versions of the tools that we use on a per-commit basis.
+
+### Install
```
These two headings visually in the same level:
![image](https://user-images.githubusercontent.com/2946214/77385441-11a07480-6dc3-11ea-9fa5-5b735c400992.png)

I thought `### Install` is a subsection of `### (optional) asdf`?

felixfbecker commented on dev/start.sh:
```diff
@@ -60,6 +60,10 @@ export ZOEKT_HOST=localhost:3070
 
+# Caddy / HTTPS configuration
+export SOURCEGRAPH_HTTPS_DOMAIN=sourcegraph.test
+export SOURCEGRAPH_HTTPS_PORT=3443
```
Sorry if I wasn't clear, this is what I had in mind so it is overridable
```suggestion
export SOURCEGRAPH_HTTPS_DOMAIN=${$SOURCEGRAPH_HTTPS_DOMAIN:-sourcegraph.test}
export SOURCEGRAPH_HTTPS_PORT=${$SOURCEGRAPH_HTTPS_PORT:-3443}
```

felixfbecker commented on doc/dev/local_development.md:
````diff
@@ -252,7 +303,59 @@ eval $(docker-machine env)
+its keys to your system's certificate store. You can get this out the way after installing Caddy 2 by running the following command and entering your password when prompted:
+
+```bash
+> env SOURCEGRAPH_HTTPS_DOMAIN=sourcegraph.test SOURCEGRAPH_HTTPS_PORT=3443 ./dev/caddy run --watch --config=dev/Caddyfile
````
Just noticed `./dev/caddy` won't work
```suggestion
env SOURCEGRAPH_HTTPS_DOMAIN=sourcegraph.test SOURCEGRAPH_HTTPS_PORT=3443 ./dev/caddy.sh run --watch --config=dev/Caddyfile
```

ggilmore commented:
@felixfbecker anything else you noticed?
